cardhost/app/lib/builder.ts

```diff
@@ -11,12 +12,35 @@ import { setComponentTemplate } from '@ember/component';
 export default class Builder implements BuilderInterface {
   private compiler = new Compiler({
     lookup: (url: string) => this.getCompiledCard(url),
+    define: (cardURL, module, source) =>
+      this.defineModule(cardURL, module, source),
   });
 
   private cache: Map<string, CompiledCard>;
+  private modulePrefix: string;
+  customDefine: (fullModulePath: string, source: unknown) => void;
 
-  constructor() {
+  constructor(params: {
+    modulePrefix?: string;
+    defineModule: (fullModulePath: string, source: unknown) => void;
+  }) {
     this.cache = new Map();
+    this.customDefine = params.defineModule;
+    this.modulePrefix = params.modulePrefix ?? '';
+  }
+
+  // TODO: is there a way to reuse the arguments of the defineModuleCallback type?
```

Comment:
Yes:

```ts
async defineModule(...args: Parameters<defineModuleCallback>) {
```

`Parameters` is a type that ships with TypeScript itself. If you use VSCode to jump-to-definition on it, you can see a whole family of similar helpers, and their (very short) implementations. In this case the implementation:

```ts
/**
 * Obtain the parameters of a function type in a tuple
 */
type Parameters<T extends (...args: any) => any> = T extends (...args: infer P) => any ? P : never;
```

is using `infer`, a pretty neat feature that's explained [here](https://www.typescriptlang.org/docs/handbook/2/conditional-types.html#inferring-within-conditional-types). Background reading for that section if it's totally alien would be starting at [this chapter](https://www.typescriptlang.org/docs/handbook/2/types-from-types.html).